### PR TITLE
Check ContextVar in default_client

### PR DESCRIPTION
**Pull request by mrocklin**
Previously default_client would iterate through the list of clients for the most recent.  Now it checks the ContextVar first.  This means that default_client will be more thread-aware.

This is particularly helpful when using `wait` in a context with multiple clients in multiple threads, a workflow that shows up in large scale XGBoost workloads.

Also, I noticed that mypy was sad with this change, which I wouldn't quite understand.  This didn't pass pre-commit on my machine.

cc @dankerrigan

### CODE REVIEW

**Comment by fjetter** [comment]
Can you add a test for this, please? The entire default client mechanism is pretty fragile. A test would also help to understand where this actually matters and why.

Mypy complains about it because the contextvar is untyped and mypy infers its type from the default, which is None.

adding annotations to the contextvar should resolve the problem
```python
_current_client: ContextVar[Client | None] = ContextVar("_current_client", default=None)
```

File: distributed/tests/test_client.py [hunk 1]
```diff
@@ -4095,6 +4095,7 @@ def run():
                     # This line runs only when all parties are inside the
                     # context manager
                     assert Client.current(allow_global=False) is c
+                    assert default_client() is c
```
Comment (mrocklin) [inline distributed/tests/test_client.py:4098]: @fjetter does this test work for you?  I've extended the existing test for `Client.current` to also run the same test for `default_client`

**Comment by mrocklin** [comment]
@fjetter I included the dask/dask PR in CI here.  Everything seems mostly green so far (although I'm not sure about the subprocess leak failure).

**Comment by mrocklin** [comment]
@fjetter thoughts on this?

I'm getting this error once in CI on Windows

    FAILED distributed/tests/test_failed_workers.py::test_submit_after_failed_worker_async[False] - AssertionError: Test leaked subprocesses: [<SpawnProcess name='Dask Worker process (from Nanny)' pid=2124 parent=6472 started daemon>]

**Comment by fjetter** [comment]
> I'm getting this error once in CI on Windows

Unfortunately very common

**Comment by fjetter** [comment]
IIUC this effectively makes `default_client` identical to `Client.current(allow_global=True)`

1. Always prefer current client
2. Fall back to *a* global client

I think this is a good step. I don't necessarily love the ambiguity of having two API calls but I can live with that.

File: .github/workflows/tests.yaml [hunk 2]
```diff
@@ -107,6 +107,11 @@ jobs:
         run: mamba env update -n dask-distributed -f ${{ env.CONDA_FILE }}
         if: steps.skip-caching.outputs.trigger-found == 'true' || steps.cache.outputs.cache-hit != 'true'
 
+      - name: Install Dev Dask
+        shell: bash -l {0}
+        run: |
+          python -m pip install --no-deps git+https://github.com/mrocklin/dask@default-client-scheduler
+
```
Comment (fjetter) [inline .github/workflows/tests.yaml:114]:
```suggestion
```

File: distributed/tests/test_client.py (hunk 1 above)
Comment (fjetter) [reply distributed/tests/test_client.py:4098]: yes. That works

**Review by fjetter — APPROVED** [review]

**Review by fjetter — APPROVED** [review]

**Pull request merged**